### DISCUSSION TITLE
v0.90 Important Announcement

### DISCUSSION THREAD

**Discussion by wingkwong** [nocodb/nocodb#3418] [Announcements]
# IMPORTANT ANNOUNCEMENT

On behalf of NocoDB Team, I would like to announce that we have made a significant number of new features and improvements and will be released as `v0.90` soon. However, some breaking changes are expected from `v0.90` onwards.

UPD: [v0.90 Release Note](https://github.com/nocodb/nocodb/releases/tag/0.90.0)

### REST APIs

- The URL paths to APIs are changing and the old URL paths **will be deprecated on 30 Jun 2022**.
- The responses to APIs will be changed.
- Swagger UI will be removed and may be supported in the future release.

| Old URL Path (< v0.90)                                   | New URL Path (v0.90)                                                                                  |
|---------------------------------------------|------------------------------------------------------------------------------------------|
| /nc/{projectId}/api/v1/{tableAlias}<br>`Deprecated on 30 Jun 22`         | /api/v1/db/data/{orgs}/{projectName}/{tableName} <br /> /api/v1/db/data/{orgs}/{projectName}/{tableId}              |
| /nc/{projectId}/api/v1/{tableAlias}/{rowId}<br>`Deprecated on 30 Jun 22` |/api/v1/db/data/{orgs}/{projectName}/{tableName}/{rowId}                                           |

> Currently, the default value for {orgs} is `noco`. Users will be able to change it in the future release.

### GraphQL APIs
GraphQL APIs, unfortunately, will be deprecated form `v0.90` onwards - which means 
- Users won't be able to create a GraphQL project nor use the GraphQL queries. 
- For projects created before `v0.90`, we will convert it to REST API projects. 
- The rationale behind is that GraphQL is a really small use case of NocoDB users and smart spreadsheet are usually flat representation of data (may be one level nested) which means there will be additional wrangling of json and GQL schema when users are changing the schema dynamically. 

## Thank You

We would like to thank all the users and contributors to join the journey with us. If you have any questions, please feel free to post your questions to [Discord](https://discord.gg/5RgZmkW). If you are interested to contribute, please check out [Contribution Guide](https://github.com/nocodb/nocodb/blob/master/.github/CONTRIBUTING.md).

**Comment by adarshmadrecha** [comment]
I am just wondering, when we have "Discussions" enabled for the repo, shouldn't these announcements be posted in "Discussions"? in "Announcement" Category?

No complaints. You guys are building amazing product. I personally think other users will be looking for these kinds of announcements in ["Discussions"](https://github.com/nocodb/nocodb/discussions)

**Comment by j2l** [comment]
Hi,
It seems so promising but you lost my business at "GraphQL APIs, unfortunately, will be deprecated form v0.90 onwards".
So sad to see a well crafted GraphQL tool for m2m tables going away.
Sure it was not fully documented, but once you got how it works, NocoDB became so powerful! Far beyond Airtables.

Next release I won't upgrade because it'll be limited to REST, competing with all other solutions like postgREST, that can already deal for delivering REST API for pg (my use case) and already implementing lots of features, even m2m (but it's hard to craft).

I really hope for a fork keeping at least GraphQL working as it is right now.

**Comment by remcoboerma** [comment]
Agrees with @j2l ; GraphQL and swagger were two reasons to start using nocodb; hope they will return (soon). 
My main reason for not investing in nocodb yet was because more difficult schema's weren't yet supported (as discussed on discord). As soon as nocodb can handle more complexity, it would be absolutely great to use it's graphql support. Removing graphql because it's not used that much seems like a self-fulfilling promise to me.

**Comment by o1lab** [comment]
@remcoboerma You can find [all apis](https://all-apis.nocodb.com) here and we've users who are appreciating these docs. And if you are interested in only data apis - [here is the link](https://apis.nocodb.com). Swagger apis for each project will be brought back. 

With >0.90 version - you should be able to do nested lookups, rollups etc (if by complex schema you meant that)

Re GraphQL : if you could expand your usecase with database type - that will helpful. We can consider bringing it back given all the inputs. But please do consider its a significant effort on our side.

**Comment by robinvandernoord** [comment]
@o1lab 
> You can find [all apis](https://all-apis.nocodb.com/) here and we've users who are appreciating these docs

I personally find these new docs very confusing compared to the old swagger docs and it now takes a lot longer to get started with the API. I know there is an SDK available but not everyone uses node, so having good documentation is important for the Dev Experience.

**Comment by o1lab** [comment]
@robinvandernoord : We hear you. Do note I did mention "Swagger apis for each project will be brought back." :) and issue is tracked [here](https://github.com/nocodb/nocodb/issues/1769).

**Comment by remcoboerma** [comment]
> @remcoboerma You can find [all apis](https://all-apis.nocodb.com) here and we've users who are appreciating these docs. And if you are interested in only data apis - [here is the link](https://apis.nocodb.com). Swagger apis for each project will be brought back. 

Glad to hear, like @robinvandernoord i think swagger makes things a lot easier, which is good for lowering the learning curve. 

> With >0.90 version - you should be able to do nested lookups, rollups etc (if by complex schema you meant that)

I'm eager to get my hands on that and see what I can do. Thanks for the collective development effort 

> Re GraphQL : if you could expand your usecase with database type - that will helpful. We can consider bringing it back given all the inputs. But please do consider its a significant effort on our side.

I can imagine it being a lot of work. Most things (if not everything) graphql related seems like a lot of work, but it can result in high quality apis. 

I'll write up a summary, where should I post it? 
If generated openapi docs contain the complete structure and metadata, it might even be an option so seperate/extract the graphql processor from nocodb:

When Nocodb produces the Openapi schema, which is fed to a graphql-processor-generator which produces a service that translates the graphql queries and actions to rest verbs on the openapi schema... It would require a rebuild on every schema change, but that should be doable using (web)hooks. All of that beauty could be in a seperate code base and be even more general to use and might even be developed outside the scope of the nocodb team. Does this sound to far fetched? 

Regards

**Comment by o1lab** [comment]
> where should I post it?

Our github discussion will be good.

> It would require a rebuild on every schema change

This rebuilding can get a bit tricky since :
1. Its not just rebuilding on table or column or link -> add/edit/delete/rename etc
2. But also bringing nested rollups, lookup, formulas to root level (this doesnt go well with gql)
3. And then we have REST API for each seperated view within a table, for which there will be new types in gql that will have to generated. We've users who have 50 to 100+ views within a table. At this point - having gql schema doesn't make much sense. Plus you would have keep track of all these additional sub types generated when step 1,2 happens :)

However, we can easily put back a simple gql just for the models alone. But it wouldnt be matching the spreadsheet

**Comment by zhzy0077** [comment]
How can I re-enable old URLs on 0.9 versions? It's being said "will be deprecated" rather than "has been removed"

**Comment by wingkwong** [comment]
@zhzy0077 The old APIs are still here for time-being and you can use them via REST API calls, but not from GUI as it is using the new ones. After deprecation, the old APIs will be completely removed.

**Comment by zhzy0077** [comment]
@wingkwong Unfurtunately no, I see `Cannot GET /nc/demo_0r5n/api/v1/demo` once I upgraded to [0.90.5]
Simple repro:
1. Provision 0.84.15
1. Create a project demo
2. Create a table demo
3. Call `http://localhost:8080/nc/demo_0r5n/api/v1/demo` and it works.
4. Upgrade to 0.90.5
5. I see 404 when calling it.

It's also reported elsewhere at https://github.com/n8n-io/n8n/issues/3170

It looks like they're all named V2 now. I don't know why https://github.com/nocodb/nocodb/blob/develop/packages/nocodb/src/lib/noco/meta/api/dataApis/oldDataApis.ts#L145

**Comment by zhzy0077** [comment]
@wingkwong By the way, even on 0.90.7 which has the fix #1831, the old URL returns a totally different data structure.
0.84.15:
```json
[
    {
        "id": 1,
        "title": "test",
        "created_at": "2022-04-24 04:44:35",
        "updated_at": "2022-04-24 04:44:35"
    }
]
```
0.90.7
```json
{
    "list": [
        {
            "Id": 1,
            "Title": "123",
            "CreatedAt": "2022-04-24 04:30:39",
            "UpdatedAt": "2022-04-24 04:30:39"
        }
    ],
    "pageInfo": {
        "totalRows": 1,
        "page": 1,
        "pageSize": 25,
        "isFirstPage": true,
        "isLastPage": true
    }
}
```
This makes a little sense for backwards compatibility as old comsumers won't work at all.

**Comment by robinvandernoord** [comment]
Sorry I misread something, removing comment :)

Btw thanks for re-adding swagger!

**Comment by zhzy0077** [comment]
@robinvandernoord Everything you said makes sense to me. I would suggest directly remove old paths if we return a totally different data structure.
1. It leads to higher troubleshooting effort for clients. A null pointer exception is much harder to investigate than a 404.
2. The old apis will be deprecated sometime later and it's not backwards compatible, it introduces efforts to maintain them and no benifits at all.

At least, we should explicitly mention that the structure is changed since it's a breaking change.

**Comment by wingkwong** [comment]
Locking the convo. For discussions related to `0.90.x`, please move to [0.90.x Discussion](https://github.com/nocodb/nocodb/discussions/1897) instead.
